docs/julia.rst

```diff
@@ -0,0 +1,450 @@
+number of return values of the function called. An expression that is also
+a statement is invalid if it evaluates to more than one value, i.e. at the
+block-level, only expressions evaluating to zero values are allowed.
+
+For variable declarations and assignments, the right-hand-side expression
+(if present) has to evaluate to a number of values equal to the number of
+variables on the left-hand-side.
+
+An expression used as an argument to a function call has to evaluate to exactly
+one value.
+
+The ``continue`` and ``break`` statements can only be used inside loop bodies.
+The condition part of the for-loop has to evaluate to exactly one value.
+
+Literals cannot be larger than the their type. The largest type defined is 256-bit wide.
+
+Formal Specification
+--------------------
+
+We formally specify JULIA by providing an evaluation function E overloaded
+on the various nodes of the AST. Any functions can have side effects, so
+E takes a state objects and the actual argument and also returns new
+state objects and new arguments. There is a global state object
+(which in the context of the EVM is the memory, storage and state of the
+blockchain) and a local state object (the state of local variables, i.e. a
+segment of the stack in the EVM).
+
+The the evaluation function E takes a global state, a local state and
+a node of the AST and returns a new global state, a new local state
+and a variable number of values.
+
+The exact nature of the global state is unspecified for this high level
+description. The local state `L` is a mapping of identifiers `i` to values `v`,
+denoted as `L[i] = v`.
+The special value `⊥` is used to signify that a variable cannot be
+used yet.
+
+.. code::
+    E(G, L, <{St1, ..., Stn}>: Block) =
+        let L' be an extension of L to all variables v declared in Block
+        (but not in its sub-blocks), such that L'[v] = ⊥.
+        let Gi, Li, mode = E(G, L', St1, ..., Stn)
+        let L'' be a restriction of Li to the identifiers of L
+        Gi, L'', mode
+    E(G, L, St1, ..., Stn: Statement) =
+        if n is zero:
+            G, L
+        else:
+            let G', L', mode = E(G, L, St1)
+            if mode is regular then
+                E(G', L', St2, ..., Stn)
+            otherwise
+                G', L', mode
+    E(G, L, <function fname (param1, ..., paramn) -> (ret1, ..., retm) block>: FunctionDefinition) =
+        G, L, regular
+    E(G, L, <let var1, ..., varn := rhs>: VariableDeclaration) =
+        E(G, L, <var1, ..., varn := rhs>: Assignment)
+    E(G, L, <let var1, ..., varn>: VariableDeclaration) =
+        let L' be a copy of L where L'[vi] = 0 for i = 1, ..., n
+        G, L', regular
+    E(G, L, <var1, ..., varn := rhs>: Assignment) =
+        let G', L', v1, ..., vn = E(G, L, rhs)
+        let L'' be a copy of L' where L'[vi] = vi for i = 1, ..., n
+        G, L'', regular
+    E(G, L, <for { i1, ..., in } condition post body>: ForLoop) =
```

Comment:
These are the statements inside a block. We can also leave out the commas, not sure how to write that best. Perhaps it is better to use `init` here and then explode it into statements in the next line.